lib/src/tree.rs

```diff
@@ -322,15 +320,25 @@ impl<T> Diff<T> {
 }
 
 struct TreeEntryDiffIterator<'trees> {
-    it1: Peekable<TreeEntriesNonRecursiveIterator<'trees>>,
-    it2: Peekable<TreeEntriesNonRecursiveIterator<'trees>>,
+    tree1: &'trees Tree,
+    tree2: &'trees Tree,
+    basenames: Vec<&'trees RepoPathComponent>,
```

Comment:
nit: this could be `vec::IntoIter<>` (to eliminate `.reverse()`) or `Box<dyn Iterator<>>` (to not `.collect_vec()`), but I don't know if that would matter.

Reply:
I don't think `vec::IntoIter<>` would help because these are just refs, so copying them should be cheap. Switching to `Box<dyn Iterator>` doesn't seem to have any impact on performance, but it's slightly simpler, and it should at least make it slightly cheaper to abort iteration early while visiting a large directory, if we ever have a use case for that (i.e. better best-case performance, and seemingly equal worse-case performance).